Kieran/celeb crud

kieranLacey opened the pull request:
Completed CRUD for Celeb including routes, views and testing for both

kieranLacey committed:
Add celebs.create view

kieranLacey committed:
Merge branch 'main' into kieran/celebCrud

kieranLacey committed:
Fixed some testing errors after merging main into branch

kieranLacey committed:
Merge branch 'main' into kieran/celebCrud

# Conflicts:
#	database/seeders/DatabaseSeeder.php
#	resources/views/landing.blade.php

kieranLacey committed:
edited celebs.create, commiting just to check something on main

kieranLacey committed:
Merge branch 'main' into kieran/celebCrud

# Conflicts:
#	routes/web.php

kieranLacey committed:
Added request validation for Celeb and started on the logic for adding a celeb to the database

kieranLacey committed:
Completed celeb create functionality and implemented the view for celeb edit functionality

kieranLacey committed:
Added testing for admin on celeb for destroy, and update as well as validation testing for celeb.store and celeb.edit, started linking the views for edit and create

kieranLacey committed:
Fixed up some visual issues in celeb.index, celeb.show, home and app

kieranLacey committed:
Added extra syntax needed for errors and changed some visuals on the Add Celeb button

vrodriguero commented on app/Http/Controllers/CelebController.php:
```diff
@@ -22,14 +23,16 @@ public function create()
-        //
+        Celeb::create($request->validated());
+
+        return redirect('celebs');
```
You could probably have a success message appear after you redirect to the celebs. index.

vrodriguero commented on app/Http/Controllers/CelebController.php:
```diff
@@ -22,14 +23,16 @@ public function create()
     public function show(Celeb $celeb)
     {
-        $celeb->with('movies')->get();
+        $celeb->with('movies')->paginate(5);
```
Hey Kieran,

Nice idea to paginate the movies. However, if you check the view it does not seem to be working as expected. 
Try to look around for options to paginate relationships in Laravel, you might need to store this in another variable :D and call `movies()` instead of `with`. 

![Screen Shot 2021-04-29 at 10 49 49 am](https://user-images.githubusercontent.com/53244381/116489572-a172f500-a8d8-11eb-8afe-0dbf059f7937.png)

vrodriguero commented on app/Http/Controllers/CelebController.php:
```diff
@@ -41,13 +44,17 @@ public function edit(Celeb $celeb)
-        //
+        $celeb->update($request->validated());
+
+        return redirect("celebs/$celeb->id");
```
Perhaps also redirect with a success message?

vrodriguero commented on app/Http/Requests/CelebRequest.php:
```diff
@@ -0,0 +1,35 @@
+    public function rules()
+    {
+        return [
+            'name' => [
+                'required',
+                'string',
+            ],
+            'date_of_birth' => ['required'],
```
I think it would be a good idea to add a size limitation on the name and validation to see if the date of birth is less than the current date.

vrodriguero commented on package-lock.json:
```diff
@@ -1,15357 +1,6 @@
-            }
-        }
-    },
+    "lockfileVersion": 1,
```
Unless you have not added any packages, I think this should not change. You can revert this to be the same as the main.

vrodriguero commented on resources/views/celebs/show.blade.php:
```diff
@@ -2,24 +2,34 @@
+                        <span class="flex">{{$title->pivot->character_name}}</span>
+                        <span class=" flex italic">{{$title->title}}</span>
+                    </div>
+                @endforeach
```
As you will paginate the movies, I think you should have the pagination links here?

vrodriguero commented on resources/views/home.blade.php:
```diff
@@ -96,7 +98,7 @@
                 <div class="grid gap-8 mt-8 sm:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4">
                     @foreach ($celebs as $celeb)
-                        <a href={{ url("celebs/$celeb->id") }}>
+                        <a href={{ route("celebs.show", $celeb->id) }}>
```
Nice!

vrodriguero commented on tests/Http/MovieTest.php:
```diff
@@ -14,7 +14,8 @@ class MovieTest extends TestCase
     public function an_admin_can_see_movies_view()
     {
-        $admin = User::factory()->admin()->create();
+        $this->withoutExceptionHandling();
```
You can delete `$this->withoutExceptionHandling()`

vrodriguero reviewed (changes requested):
Hey Kieran!

Great job on the challenges! 😎 

I have added some suggestions for this PR ;D

kieranLacey committed:
Starting to implement pagination for movie titles on celebs.show

kieranLacey committed:
Finished logic for pagination on celeb.show, working on implementing pagination visuals into current celeb.show view

kieranLacey committed:
Added pagination for celebs.show and added success messages for adding, updating and deleting celebs

kieranLacey committed:
Changes after merging main into branch

kieranLacey committed:
Edited some visual aspects of the celeb.show view

kieranLacey committed:
Reverted package-lock.json to how it is in main

vrodriguero committed:
Merge branch 'main' into kieran/celebCrud

# Conflicts:
#	routes/web.php

vrodriguero committed:
wip:
- Added check for celeb request
- added admin middleware on Kernel
- update tests for celeb, after middleware change

vrodriguero committed:
wip:
- created a simple admin middleware test
- Moved celebs test to Http
- Updated tests names
- Fix namespaces for celeb and movie tests

vrodriguero committed:
wip:
- fix movie test

vrodriguero committed:
wip

vrodriguero commented on tests/Http/CelebTest.php:
```diff
@@ -0,0 +1,241 @@
+        $celeb = \App\Models\Celeb::factory()->create();
+        $this->deleteJson("celebs/$celeb->id")
+            ->assertRedirect();
+    }
```
I have made some changes to the names of the tests Kieran, but I notice that are some tests that are worth doing it. For example, testing if a user inputs a date of birth after today's date and if the user tries to add a name that is longer or smaller than the allowed number. In this case, you would want to test if the message is as you expect :D not only if the session has that error key 😄

vrodriguero reviewed (changes requested):
All done Kieran,

Please note that I have made some change to the Celebs tests such as:
- Moved the celeb tests to the HTTP and in the Middleware, I actually created a generic test for the middleware. So moving forward I think this structure will be better to find tests, because before the tests for the AdminCeleb, was actually testing much more HTTP stuff. 😄 
- rename the test to CelebTest. 

Just check the comment regarding some missing tests, which I think it should be implemented 😄 

Great work!

vrodriguero committed:
wip:
- Moved celeb and movie test to controllers

vrodriguero committed:
changes to resolve conflict

vrodriguero committed:
Merge branch 'main' into kieran/celebCrud

# Conflicts:
#	tests/Http/Controllers/CelebTest.php

vrodriguero committed:
fix conflicts

kieranLacey committed:
Fixing testing errors in MovieTest

kieranLacey committed:
Merge branch 'kieran/celebCrud' of github.com:codinglabsau/moviedeck into kieran/celebCrud

kieranLacey committed:
Added testing for celeb name field being too short or too long and for date_of_birth being after todays date

kieranLacey committed:
Added testing for celeb update for name field being too short or too long and date_of_birth field being after todays date

vrodriguero commented on tests/Http/MovieTest.php:
```diff
@@ -59,23 +59,27 @@ public function a_user_cannot_access_create_movie_view()
     /** @test */
     public function an_admin_can_add_a_movie()
     {
+        $this->withoutExceptionHandling();
```
You can remove this line ;D.

vrodriguero commented on database/seeders/DatabaseSeeder.php:
```diff
@@ -25,7 +25,12 @@ public function run()
         \App\Models\Celeb::factory(100)->create();
         \App\Models\Watchlist::factory(15)->create();
-        \App\Models\CelebMovie::factory(200)->create();
+        \App\Models\CelebMovie::factory(500)->create();
+        \App\Models\CelebMovie::factory(100)->create([
+            'celeb_id' => 1,
+            'movie_id' => 1,
+            'character_name' => 'Kelsier'
+        ]);
```
Why this duplicate?

kieranLacey replied:
That was just so I could test the pagination views if it got more than 5 pages of titles for a celeb

vrodriguero replied:
Hmm... probably we can either remove this and extract it into a seeder or change the factory? That we can remove the duplicates?

vrodriguero commented on tests/Http/Middleware/AdminMiddlewareTest.php:
```diff
@@ -31,7 +31,7 @@ public function non_admin_are_redirected()
 
     /** @test */
-    public function admin_are_redirected()
+    public function admin_can_access_create_page_redirected()
```
This one is a generic test so that is why in the set up I created a generic route. You do not need to include the celeb route in here 😄

vrodriguero replied:
I have updated the main so this should be now fixed

vrodriguero commented on app/Http/Requests/CelebRequest.php:
```diff
@@ -0,0 +1,39 @@
+     */
+    public function authorize()
+    {
+        return auth()->check() && auth()->user()->is_admin;
```
This one should check if the is_admin is === true :D otherwise it would not matter if it is false or not

vrodriguero commented on tests/Http/Controllers/CelebTest.php:
```diff
@@ -13,164 +14,334 @@ class CelebTest extends TestCase
         $this->actingAs($user)
             ->putJson("celebs/$celeb->id")
-            ->assertStatus(403);
+            ->assertRedirect();
```
For this one, what do you also want to check is if the database results did not change.

So first change the $celeb to create something with the results you want 
` $celeb = Caleb::factory()->create(['name' =>  something ])` Do that for all the fillable columns
then in the put, you should change the value of one of them. 

Finally, at the end of the test, you should assert the database has the initial values and did not change. 

This is the rule of thumb for all tests that involve changes to the DB. So DELETE and CREATE also should be like that.

vrodriguero commented on tests/Http/Controllers/CelebTest.php:
```diff
@@ -13,164 +14,334 @@ class CelebTest extends TestCase
+        $this->actingAs($admin)
+            ->put("celebs/$celeb->id", [
+                'name' => 'James Islington',
+                'date_of_birth' => '08/03/2999',
```
To best check this you should use Carbon, we usually avoid typed dates on tests. 

When using Carbon we also want to make sure to set up Carbon now(). To do that you have to add a setUp() method, like the example below:
```
 protected function setUp(): void
    {
        parent::setUp();

        Carbon::setTestNow(now());
    }
```

Then in your test you can use the Carbon::now()->addDay() or whatever function that you need you can check [Carbon docs ](https://carbon.nesbot.com/docs/)

vrodriguero commented on tests/Http/Controllers/CelebTest.php:
```diff
@@ -13,164 +14,334 @@ class CelebTest extends TestCase
+        $admin = User::factory()->admin()->create();
+        $celeb = Celeb::factory()->create();
 
         $this->actingAs($admin)
             ->deleteJson("celebs/$celeb->id")
-            ->assertOk();
+            ->assertStatus(302);
+
+        $this->assertDatabaseMissing('celebs', [
+            'id' => $celeb->id,
+        ]);
     }
 
     /** @test */
-    public function normal_user_cannot_access_celebs_destroy()
+    public function user_cannot_delete_a_celeb()
     {
-        $celeb = \App\Models\Celeb::factory()->create();
-        $user = \App\Models\User::factory()->create([
+        $celeb = Celeb::factory()->create();
+        $user = User::factory()->create([
             'is_admin' => false
         ]);
 
         $this->actingAs($user)
             ->deleteJson("celebs/$celeb->id")
-            ->assertStatus(403);
+            ->assertRedirect();
     }
 
     /** @test */
-    public function guest_cannot_access_celebs_destroy()
+    public function guest_cannot_delete_a_celeb()
     {
-        $celeb = \App\Models\Celeb::factory()->create();
+        $celeb = Celeb::factory()->create();
         $this->deleteJson("celebs/$celeb->id")
-            ->assertStatus(403);
+            ->assertRedirect();
     }
```
same comment as the one in the user_cannot_update_celeb

vrodriguero commented on tests/Http/Controllers/CelebTest.php:
```diff
@@ -13,164 +14,334 @@ class CelebTest extends TestCase
+    }
+
+    /** @test */
+    public function validation_error_when_admin_updates_celeb_with_name_being_too_short()
```
I noticed that the validation is being run for both create and update right? This is okay, but in our case where they both share the same CelebRequest you can just use one endpoint, so you can rename it to something like `validation_errors_when_celeb_name_length_is_invalid` and just choose one of the endpoints. Does that make sense?

vrodriguero reviewed (changes requested):
Hi Kieran, just a couple of notes on your tests, please make sure that you read through them and let me know if you have any questions ;D

kieranLacey committed:
Merge branch 'main' into kieran/celebCrud

# Conflicts:
#	tests/Http/Middleware/AdminMiddlewareTest.php

kieranLacey committed:
Commit to merge main into branch

kieranLacey committed:
Merge branch 'main' into kieran/celebCrud

kieranLacey committed:
Editing update and delete testing for celeb so that they test if the database has the correct data or not

kieranLacey committed:
Adding Carbon for years in CelebTest

kieranLacey committed:
Merge branch 'main' into kieran/celebCrud

kieranLacey committed:
Removing the === true from admin conditionals

vrodriguero commented on tests/Http/Controllers/CelebTest.php:
```diff
@@ -2,175 +2,414 @@
+        $admin = User::factory()->admin()->create();
+
+        $this->actingAs($admin)
+            ->post('celebs', [
+                'name' => null,
+                'date_of_birth' => null,
+                'photo' => null
```
This one you can also simplify by just posting an empty array 😄 

And in that case, we do not need to check if the DB has the object because we are just checking validation error.

vrodriguero replied:
That is the same for all tests that you are only concern with the validation error, and also because we are checking update and store in other tests, so no need to worry about the DB in the validation tests.

vrodriguero commented on tests/Http/Controllers/CelebTest.php:
```diff
@@ -2,175 +2,414 @@
+                'name' => 'Jennifer Jaxon',
+                'date_of_birth' => Carbon::now()->subYears(30),
+                'photo' => 'https://generic/photo1_700x600'
+            ])->assertSessionHasNoErrors();
```
I believe this should assert ok instead of assert session has no errors.

kieranLacey replied:
Hmm this was giving me a 302 status code apparently (I have now changed it to assertStatus(302))

vrodriguero commented on tests/Http/Controllers/CelebTest.php:
```diff
@@ -2,175 +2,414 @@
+                'photo' => 'https://genericPhotos.com'
+            ])->assertRedirect();
+
+        $this->assertDatabaseMissing('celebs', [
+            'id' => $celeb->id,
+            'name' => 'Joe Invalid',
+            'date_of_birth' => Carbon::now()->subYears(30),
+            'photo' => 'https://genericPhotos.com'
+        ]);
```
You can also just leave the one below, if we are checking that the DB has the original one, we do not need to check if the record you send is missing 😄

vrodriguero reviewed (changes requested):
Okay Kieran 😄 
Just a really small changes ;D

kieranLacey committed:
Ediited testing on celeb

vrodriguero reviewed (approved):
Great work :D
